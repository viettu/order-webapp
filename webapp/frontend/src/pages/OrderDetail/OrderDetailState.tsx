import React from 'react';
import { useMutation, useQuery } from '@apollo/client';
import { Box, Button, HStack, Spinner, Text } from '@chakra-ui/react';
import { CANCEL_ORDER_MUTATION, OrderStates, QUERY_ORDER_STATE } from '../../data';
import { useEffect } from 'react';
import { OrderStateBadge } from '../../components';

type OrderDetailStateProps = {
  orderId: number;
  pollingStatusInterval?: number;
};

const OrderDetailState: React.FC<OrderDetailStateProps> = ({ orderId, pollingStatusInterval }) => {
  const { data, stopPolling, error } = useQuery(QUERY_ORDER_STATE, {
    variables: { id: orderId },
    pollInterval: pollingStatusInterval,
    fetchPolicy: 'network-only',
  });

  const [cancelOrder] = useMutation(CANCEL_ORDER_MUTATION, {
    variables: { id: orderId },
  });

  const state = data?.getOrderState;
  useEffect(() => {
    if (state === OrderStates.CANCELLED || state === OrderStates.DELIVERED) {
      stopPolling();
    }
  }, [state, stopPolling]);

  if (error) {
    return null;
  }

  return (
    <>
      <Box>
        <OrderStateBadge state={state} fontSize={'1.2em'}></OrderStateBadge>
        {state === 'CONFIRMED' && (
          <Button onClick={() => cancelOrder()} bgColor={'red.300'} size={'sm'} ml={10}>
            Cancel order
          </Button>
        )}
      </Box>

      {state === 'CONFIRMED' && (
        <HStack spacing={'10px'} alignItems={'center'} mt={5}>
          <Spinner size="md" color="teal" />
          <Text as="span" color={'teal'} fontSize={'0.9em'}>
            Your order takes 10 seconds for delivering. By clicking "Cancel order" you will cancel your order
            immediately.
          </Text>
        </HStack>
      )}
    </>
  );
};

OrderDetailState.defaultProps = {
  pollingStatusInterval: 0,
};

export default OrderDetailState;
