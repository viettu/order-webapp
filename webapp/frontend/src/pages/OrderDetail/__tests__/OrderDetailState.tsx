import React from 'react';
import OrderDetailState from '../OrderDetailState';
import renderer from 'react-test-renderer';
import { TestComponentWrapper } from '../../../utils';

test('snapshot', () => {
  const component = renderer.create(
    <TestComponentWrapper>
      <OrderDetailState orderId={1}></OrderDetailState>
    </TestComponentWrapper>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
