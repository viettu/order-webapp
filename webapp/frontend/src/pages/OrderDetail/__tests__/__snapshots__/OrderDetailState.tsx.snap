// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Array [
  <div
    className="css-0"
  />,
  <span />,
  <span
    className="chakra-env"
    hidden={true}
  />,
]
`;
